File: kotlin/basics/src/main/kotlin/UnionFind.kt
package io.github.suzp1984.algorithms

class UnionFind(size : Int) {
    val ids : Array<Int>

    init {
        ids = Array(size) { it }
    }

    private fun find(p : Int) : Int {
        if (p < 0 || p > ids.size)
            throw IllegalArgumentException("array out of scope")

        return ids[p]
    }

    fun isConnected(q : Int, p : Int) : Boolean {
        if (q >= 0 && q < ids.size &&
                p >= 0 && p < ids.size) {
            return ids[p] == ids[q]
        }

        throw IllegalArgumentException("array out of scope")
    }

    fun unionElements(p : Int, q : Int) {
        val pId = find(p)
        val qId = find(q)

        if (pId == qId) {
            return
        }

        ids.indices
                .filter { ids[it] == pId }
                .forEach { ids[it] = qId }
    }
}

class UnionFind2(size : Int) {
    private val ids : Array<Int>

    init {
        ids = Array(size) { it }
    }

    private fun find(p : Int) : Int {
        if (p < 0 || p > ids.size)
            throw IllegalArgumentException("array out of scope")

        var i = p
        while (i != ids[i]) i = ids[i]

        return i
    }

    fun isConnected(q : Int, p : Int) : Boolean {
        return find(q) == find(p)
    }

    fun unionElements(p : Int, q : Int) {
        val pId = find(p)
        val qId = find(q)

        if (pId == qId) {
            return
        }

        ids[pId] = qId
    }
}

File: scala/algorithms/src/main/scala/io/github/suzp1984/sort/UnionFind.scala
package io.github.suzp1984.sort

trait UnionFind {
  def isConnected(q : Int, p: Int): Boolean
  def unionElements(q: Int, p: Int)
}

class UnionFindOne private(private val n : Int) extends UnionFind {
  val ids = new Array[Int](n)
  ids.indices.foreach { i =>
    ids(i) = i
  }

  private def find(p : Int): Int = {
    require(p >= 0 && p < n)
    ids(p)
  }

  def isConnected(q : Int, p : Int) : Boolean = {
    require(q >= 0 && q < n)
    require(p >= 0 && p < n)

    ids(p) == ids(q)
  }

  def unionElements(q : Int, p : Int): Unit = {
    val pId = find(p)
    val qId = find(q)

    if (qId == pId) {
      return
    }

    ids.indices.foreach { i =>
      if (ids(i) == pId) {
        ids(i) = qId
      }
    }
  }
}

object UnionFindOne {
  def apply(n: Int): UnionFindOne = new UnionFindOne(n)
}

class UnionFindTwo private (private val n : Int) extends UnionFind {
  val ids = new Array[Int](n)
  ids.indices.foreach { i =>
    ids(i) = i
  }

  private def find(p : Int): Int = {
    require(p >= 0 && p < n)
    var index = p
    while (index != ids(index)) {
      index = ids(index)
    }
    ids(index)
  }

  def isConnected(p: Int, q: Int) : Boolean = {
    ids(p) == ids(q)
  }

  def unionElements(q: Int, p : Int): Unit = {
    val pRoot = find(p)
    val qRoot = find(q)

    if (qRoot == pRoot) {
      return
    }

    ids(pRoot) = qRoot
  }
}

object UnionFindTwo {
  def apply(n :Int): UnionFind = new UnionFindTwo(n)
}
